### PR TITLE
scenebuilder: 15.0.1 -> 20.0.0, enable javaFX and webkit

### LINKED ISSUE #238902: scenebuilder: "Could  not find or load main class com.oracle.javafx.scenebuilder.app.SceneBuilderApp"
### Describe the bug

On current master, unstable and 23.05 HEADs, the `scenebuilder` package can be built but not started.

### Steps To Reproduce
1. `nix run nixpkgs#scenebuilder`
2. See error:
   ```text
   Error: Could not find or load main class com.oracle.javafx.scenebuilder.app.SceneBuilderApp
   Caused by: java.lang.NoClassDefFoundError: javafx/application/Application
   ```

### Expected behavior

Scenebuilder starts succesfully

### Additional context

A `git bisect` with `--first-parent` identified a `staging-next` merge from 2023-02-15 as the problem. (#216499; 30e272bc047dfa347315e9fe9bcb10a9cf1e0865)

```
30e272bc047dfa347315e9fe9bcb10a9cf1e0865 is the first bad commit
commit 30e272bc047dfa347315e9fe9bcb10a9cf1e0865
Merge: 5d1eee6cf04 1af0d67ebe1
Author: Vladimír Čunát <v@cunat.cz>
Date:   Sun Feb 26 10:11:50 2023 +0100

    Merge #216499: staging-next 2023-02-15
```

I started a more detailed bisection already though it is estimated on 8 steps, which in my experience take 6+ hours each on my hardware.

### Notify maintainers

* @wirew0rm

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-info -m
 - system: `"x86_64-linux"`
 - host os: `Linux 6.3.5, NixOS, 23.11 (Tapir), 23.11.20230617.04af42f`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.17.0pre20230619_3910430`
 - channels(nmelzer): `""`
 - channels(root): `""`
 - nixpkgs: `/etc/nixpkgs/channels/nixpkgs`
```

### PR DESCRIPTION
**Pull request by wirew0rm**
Update scenebuilder to the latest version supported by the jdk versions available in nixpkgs.

For changelogs see: https://github.com/gluonhq/scenebuilder/releases

The changes are rather big, because the project switched from gradle to maven.

This was triggered by #238902, which reported that scenebuilder broke and fails to find the javafx module on startup. There are different ways to handle the move of javafx from core jdk modules to external modules, with many distributions including nixkpgs shipping versions of the jdk which have the javafx modules added back in.
This used to be the default but got changed in https://github.com/NixOS/nixpkgs/pull/206643 / ce6bc62
to reduce the closure of non javafx java apps.

Since scenic-view has the same situation and is closely related to scene-builder concerning packaging and userbase, I fixed scenic-view in a second commit. Additionally 1caf19b from the same PR mentioned earlier also removed webkit support from openjfx which is also required for scenic-view and therefore explicitly enabled.

I've tested that the packages build correctly again and produce runnable binaries, but I didn't go into deep testing of functionalities.

fixes #238902 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by NobbZ** [comment]
Result of `nixpkgs-review pr 239254` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>scenebuilder</li>
    <li>scenic-view</li>
  </ul>
</details>

**Comment by NobbZ** [comment]
I'm worried about ofborgs eval failure, while it build locally for me…

**Comment by wirew0rm** [comment]
The ofborg error seems to be related to the darwin package, but both scenebuilder and scenic-view are marked as broken on darwin :thinking: Are the override attributes evaluated before broken is evaluated? I don't know what else I could do to make offborg not evaluate this on darwin, maybe someone with more offborg/openjdk/darwin experience can weigh in here? I'll try to investigate some more later, there's other packages using `enableJavaFx = true;` they'll have to solve or suppress the same problem somehow .

**Comment by tenaf0** [comment]
In the meanwhile openjdk20 has become packaged and merged -- I have tested @wirew0rm 's version and it seems to work fine as is. If there is interest I can help bump it to version 20.

Even if the darwin packages doesn't build, I would be happy to have it included, as the current version doesn't work at all, on any platform.

**Comment by wegank** [comment]
@ofborg build scenebuilder scenebuilder.passthru.tests

**Review by wirew0rm — COMMENTED** [review]
Thanks for taking care of this and getting it merged :+1: 
Also thanks for fixing ofborg and adding the arm support, I wouldn't have known where to start on these.

I have a few small questions on some of the changes where I'm interested in the reasoning and would be glad if you could point me in the direction of where to learn more. Also some super minor issues,that i commented on before it was merged, that I would now just follow up in the next maintainance PR.

Thanks for your work on following up and merging all those PRs by occasional contributors like myself, it's really appreciated!

File: pkgs/development/tools/scenic-view/default.nix [hunk 1]
```diff
@@ -1,5 +1,10 @@
-{ lib, stdenv, fetchFromGitHub, jdk, gradle_7, makeDesktopItem, copyDesktopItems, perl, writeText, runtimeShell, makeWrapper }:
+{ lib, stdenv, fetchFromGitHub, openjdk, openjfx, gradle_7, makeDesktopItem, perl, writeText, makeWrapper }:
 let
+  jdk = openjdk.override (lib.optionalAttrs stdenv.isLinux {
+    enableJavaFX = true;
+    openjfx = openjfx.override { withWebKit = true; };
+  });
+
```
[the remark this answers is not in the document]
Comment (wegank) [reply pkgs/development/tools/scenic-view/default.nix:7]: Personally, I'd like the JDK version to be explicit, but I suppose my change isn't necessary in terms of conventions. You can definitely revert this in the next PR.

File: pkgs/development/tools/scenic-view/default.nix [hunk 2]
```diff
@@ -1,5 +1,10 @@
-{ lib, stdenv, fetchFromGitHub, jdk, gradle_7, makeDesktopItem, copyDesktopItems, perl, writeText, runtimeShell, makeWrapper }:
+{ lib, stdenv, fetchFromGitHub, openjdk, openjfx, gradle_7, makeDesktopItem, perl, writeText, makeWrapper }:
```
[the remark this answers is not in the document]
Comment (wegank) [reply pkgs/development/tools/scenic-view/default.nix:1]: You're right, it should be added into `nativeBuildInputs`.

File: pkgs/development/tools/scenebuilder/default.nix [hunk 3]
```diff
@@ -32,7 +45,7 @@ maven.buildMavenPackage rec {
   '';
 
   postFixup = ''
-    makeWrapper ${jre}/bin/java $out/bin/${pname} \
+    makeWrapper ${jdk}/bin/java $out/bin/${pname} \
```
[the remark this answers is not in the document]
Comment (wegank) [reply pkgs/development/tools/scenebuilder/default.nix:48]: `jre` is an alias for `jdk` in nixpkgs. Perhaps you'd like to use `jre_minimal`?

[the remark this answers is not in the document]
Comment (wirew0rm) [reply pkgs/development/tools/scenebuilder/default.nix:48]: I just saw that some packages use it and it's also used in the nixpkgs manual example so i just assumed that there would be some difference. If i find the time i'll look into jre_minimal, but for a development tool it's also not that critical.

File: pkgs/development/tools/scenic-view/default.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (wirew0rm) [reply pkgs/development/tools/scenic-view/default.nix:7]: I was kind of hoping that there's just a rule that applies so I won't have to think about that in the future ;) I'll do some more research/discussing if I think about changing it back.

**Pull request merged**